presto-main/src/main/java/io/prestosql/sql/planner/PlanBuilder.java

```diff
@@ -85,28 +93,39 @@ public TranslationMap getTranslations()
         return translations;
     }
 
+    public Scope getScope()
+    {
+        return translations.getScope();
+    }
+
     public PlanBuilder appendProjections(Iterable<Expression> expressions, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
     {
-        TranslationMap translations = copyTranslations();
+        return appendProjections(expressions, symbolAllocator, idAllocator, TranslationMap::rewrite, TranslationMap::canTranslate);
+    }
 
+    public <T extends Expression> PlanBuilder appendProjections(
+            Iterable<T> expressions,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            BiFunction<TranslationMap, T, Expression> rewriter,
+            BiPredicate<TranslationMap, T> alreadyHasTranslation)
+    {
         Assignments.Builder projections = Assignments.builder();
 
         // add an identity projection for underlying plan
-        for (Symbol symbol : getRoot().getOutputSymbols()) {
-            projections.put(symbol, symbol.toSymbolReference());
-        }
-
-        ImmutableMap.Builder<Symbol, Expression> newTranslations = ImmutableMap.builder();
-        for (Expression expression : expressions) {
-            Symbol symbol = symbolAllocator.newSymbol(expression, getAnalysis().getTypeWithCoercions(expression));
-            projections.put(symbol, translations.rewrite(expression));
-            newTranslations.put(symbol, expression);
-        }
-        // Now append the new translations into the TranslationMap
-        for (Map.Entry<Symbol, Expression> entry : newTranslations.build().entrySet()) {
-            translations.put(entry.getValue(), entry.getKey());
+        projections.putIdentities(root.getOutputSymbols());
+
+        Map<ScopeAware<Expression>, Symbol> mappings = new HashMap<>();
+        for (T expression : expressions) {
+            if (!mappings.containsKey(scopeAwareKey(expression, translations.getAnalysis(), translations.getScope())) && !alreadyHasTranslation.test(translations, expression)) {
```

Comment:
`canTranslate` means that it has a mapping for the whole expression (e.g., because the expression appears multiple times). If it can't translate, we rewrite the expression (that converts all references to symbols, inserts coercions, etc), and add it to the translation map